File: VirtualPetC#/VirtualPetC#/Core/GameManager.cs
using VirtualPetC_.Core.Pets;
using VirtualPetC_.Enums;
using VirtualPetC_.UI;

namespace VirtualPetC_.Core;

/// <summary>
/// GameManager orchestrates the game flow and manages the pet lifecycle.
/// Demonstrates DEPENDENCY INVERSION PRINCIPLE - depends on IUserInterface abstraction
/// rather than concrete MenuSystem implementation.
/// </summary>
public class GameManager
{
    private BasePet? currentPet;
    private readonly IUserInterface ui;
    private bool isGameRunning;
    private DateTime lastUpdateTime;

    /// <summary>
    /// Constructor demonstrates DEPENDENCY INJECTION.
    /// GameManager depends on IUserInterface interface, not concrete implementation.
    /// This allows swapping UI implementations without changing GameManager code.
    /// </summary>
    /// <param name="userInterface">Any class that implements IUserInterface</param>
    public GameManager(IUserInterface userInterface)
    {
        ui = userInterface;
        isGameRunning = false;
        lastUpdateTime = DateTime.Now;
    }


    // Starts and runs the virtual pet simulator.
    public void Start()
    {
        ui.DisplayWelcome();
        CreateNewPet();

        if (currentPet != null)
        {
            RunGameLoop();
        }
    }


    // Creates a new pet based on user choice.
    private void CreateNewPet()
    {
        PetType petType = ui.GetPetTypeChoice();
        string petName = ui.GetPetName();

        // Create different pet types, all stored as Pet base class reference
        currentPet = petType switch
        {
            PetType.Dog => new Dog(petName),
            PetType.Cat => new Cat(petName),
            PetType.Bird => new Bird(petName),
            _ => new Dog(petName)
        };

        Console.Clear();
        Console.WriteLine($"\nYou've adopted a {petType} named {petName}!\n");

        //MakeSound() behavior depends on runtime type
        currentPet.MakeSound();

        lastUpdateTime = DateTime.Now;
        ui.PauseForUser();
    }


    /// Updates the pet's stats based on time elapsed since last update.
    private void UpdatePet()
    {
        if (currentPet == null) return;

        DateTime now = DateTime.Now;
        double deltaTime = (now - lastUpdateTime).TotalSeconds;

        currentPet.Update(deltaTime);
        lastUpdateTime = now;
    }


    /// Main game loop - continues until player exits or pet dies.
    private void RunGameLoop()
    {
        Console.Clear();
        Console.WriteLine("\x1b[3J");
        if (currentPet == null) return;

        isGameRunning = true;

        while (isGameRunning && currentPet.IsAlive)
        {
            Console.Clear();
            Console.WriteLine("\x1b[3J");
            Console.SetCursorPosition(0, 0);

            // Update pet stats based on elapsed time
            UpdatePet();

            // Show status and warnings
            currentPet.DisplayStatus();
            ui.DisplayWarnings(currentPet);
            ui.DisplaySeparator();

            // Display menu and get user choice
            ui.DisplayActionMenu(currentPet);
            int choice = ui.GetValidChoice(1, 8);

            // Execute action based on choice
            ExecuteAction(choice);

            // Check if pet died after action
            if (!currentPet.IsAlive)
            {
                Console.Clear();
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine($"\n{currentPet.Name} has died from neglect...");
                Console.ResetColor();
                ui.PauseForUser();
                break;
            }

            ui.PauseForUser();
        }

        // Game over
        ui.DisplayGameOver(currentPet);
    }

    /// <summary>
    /// Executes the player's chosen action.
    /// </summary>
    /// <param name="choice">Player's menu choice</param>
    private void ExecuteAction(int choice)
    {
        if (currentPet == null) return;

        ui.DisplaySeparator();

        switch (choice)
        {
            case 1: // Feed
                string feedResult = currentPet.Feed();
                ui.DisplayMessage(feedResult);
                break;

            case 2: // Play
                // POLYMORPHISM: Play() behavior differs for Dog, Cat, Bird
                string playResult = currentPet.Play();
                ui.DisplayMessage(playResult);
                break;

            case 3: // Sleep
                string sleepResult = currentPet.Sleep();
                ui.DisplayMessage(sleepResult);
                break;

            case 4: // Clean
                string cleanResult = currentPet.Clean();
                ui.DisplayMessage(cleanResult);
                break;

            case 5: // Interact (Make Sound)
                currentPet.MakeSound();
                break;

            case 6: // Use Special Ability
                ui.UseSpecialAbility(currentPet);
                break;

            case 7: // View Status
                Console.WriteLine("(Status already displayed above)");
                break;

            case 8: // Exit
                Console.WriteLine("\nAre you sure you want to exit? (y/n)");
                string? confirm = Console.ReadLine();
                if (confirm?.ToLower() == "y")
                {
                    isGameRunning = false;
                    Console.WriteLine($"\nGoodbye! Thanks for caring for {currentPet.Name}!");
                }
                break;

            default:
                Console.WriteLine("Invalid choice!");
                break;
        }
    }
}

File: VirtualPetGo/game/game_manager.go
package game

import (
	"VirtualPetGo/pet"
	"VirtualPetGo/ui"
	"VirtualPetGo/utils"
	"fmt"
	"time"
)

type GameManager struct {
	currentPet     pet.Pet
	lastUpdateTime time.Time
	ui             ui.IUserInterface
}

func NewGameManager(userInterface ui.IUserInterface) *GameManager {
	return &GameManager{
		currentPet:     nil,
		lastUpdateTime: time.Now(),
		ui:             userInterface,
	}
}

func (gm *GameManager) CreatePet(petType int, name string, variant string) {
	switch petType {
	case 1:
		gm.currentPet = pet.NewDog(name, variant)

	}
	gm.lastUpdateTime = time.Now()
}

func (gm *GameManager) UpdatePet() {
	if gm.currentPet == nil {
		return
	}

	//Calculate time since last Update
	now := time.Now()
	deltaTime := now.Sub(gm.lastUpdateTime).Seconds()

	gm.currentPet.Update(deltaTime)
	gm.lastUpdateTime = now
}
func (gm *GameManager) GetPet() pet.Pet {
	return gm.currentPet
}
func (gm *GameManager) Start() {
	// Display welcome screen
	gm.ui.DisplayWelcome()

	// Create the pet
	gm.createPet()

	// Run the main game loop
	gm.gameLoop()
}

// createPet handles pet creation phase
func (gm *GameManager) createPet() {
	fmt.Print("\nEnter your pet's name: ")
	name, _ := utils.ReadString()

	fmt.Print("Enter your dog's breed: ")
	breed, _ := utils.ReadString()

	// Create dog (MVP - only Dog for now)
	gm.currentPet = pet.NewDog(name, breed)
	gm.lastUpdateTime = time.Now()

	fmt.Printf("\n%s the %s has been born!\n", name, breed)
}

// gameLoop is the main game loop
func (gm *GameManager) gameLoop() {
	for {
		gm.ui.ClearScreen()
		// Update pet stats based on elapsed time
		gm.updatePet()

		// Check if pet is still alive
		if !gm.currentPet.IsAlive() {
			fmt.Printf("\n💀 %s has died... Game Over.\n", gm.currentPet.GetStatus().Name)
			break
		}
		gm.ui.DisplayStatus(gm.currentPet)
		// Display menu
		gm.ui.DisplayMainMenu()

		// Get user choice (1-8)
		choice, _ := utils.ReadIntInRange(1, 8)

		// Handle the action, returns false if user wants to exit
		if !gm.handleAction(choice) {
			fmt.Println("\nThanks for playing! Goodbye!")
			break
		}

		fmt.Print("\nPress Enter to continue...")
		utils.WaitForEnter()
		//bufio.Reader.ReadString('\n')
	}
}

// updatePet updates the pet's stats based on time elapsed
func (gm *GameManager) updatePet() {
	if gm.currentPet == nil {
		return
	}

	now := time.Now()
	deltaTime := now.Sub(gm.lastUpdateTime).Seconds()

	gm.currentPet.Update(deltaTime)
	gm.lastUpdateTime = now
}

// handleAction processes user's menu choice
// Returns false if user wants to exit, true otherwise
func (gm *GameManager) handleAction(choice int) bool {
	switch choice {
	case 1: // Feed
		result := gm.currentPet.Feed()
		gm.ui.DisplayMessage(result)

	case 2: // Play
		result := gm.currentPet.Play()
		gm.ui.DisplayMessage(result)

	case 3: // Sleep
		result := gm.currentPet.Sleep()
		gm.ui.DisplayMessage(result)

	case 4: // Clean
		result := gm.currentPet.Clean()
		gm.ui.DisplayMessage(result)

	case 5: // Interact (Make Sound)
		result := gm.currentPet.Interact()
		gm.ui.DisplayMessage(result)

	case 6: // Use Special Ability
		if gm.currentPet.CanUseAbility() {
			result := gm.currentPet.UseSpecialAbility()
			gm.ui.DisplayMessage(result)
		} else {
			gm.ui.DisplayMessage("Special ability is not available right now!")
		}

	case 7: // View Status
		gm.ui.DisplayStatus(gm.currentPet)

	case 8: // Exit Game
		return false
	}

	return true // Continue game
}
